oak/server/grpc_client_node.cc

```diff
@@ -0,0 +1,196 @@
+    return false;
+  }
+  if (invocation.msg->channels.size() != 2) {
+    LOG(ERROR) << "Wrong number of channels " << invocation.msg->channels.size()
+               << " in invocation";
+    return false;
+  }
+
+  std::unique_ptr<ChannelHalf> half0 = std::move(invocation.msg->channels[0]);
+  auto channel0 = absl::get_if<std::unique_ptr<MessageChannelReadHalf>>(half0.get());
+  if (channel0 == nullptr) {
+    LOG(ERROR) << "First channel accompanying invocation is write-direction";
+    return false;
+  }
+  std::unique_ptr<MessageChannelReadHalf> req_channel = std::move(*channel0);
+
+  std::unique_ptr<ChannelHalf> half1 = std::move(invocation.msg->channels[1]);
+  auto channel1 = absl::get_if<std::unique_ptr<MessageChannelWriteHalf>>(half1.get());
+  if (channel1 == nullptr) {
+    LOG(ERROR) << "Second channel accompanying invocation is read-direction";
+    return false;
+  }
+  std::unique_ptr<MessageChannelWriteHalf> rsp_channel = std::move(*channel1);
+
+  // Expect to read a single request out of the request channel.
+  // TODO(#97): support client-side streaming
+  ReadResult req_result = req_channel->Read(INT_MAX, INT_MAX);
+  if (req_result.required_size > 0) {
+    LOG(ERROR) << "Message size too large: " << req_result.required_size;
+    return false;
+  }
+  if (req_result.msg->channels.size() != 0) {
+    LOG(ERROR) << "Unexpectedly received channel handles in request channel";
+    return false;
+  }
+  GrpcRequest grpc_req;
+  grpc_req.ParseFromString(std::string(req_result.msg->data.data(), req_result.msg->data.size()));
+  std::string method_name = grpc_req.method_name();
+  const grpc::string& req_data = grpc_req.req_msg().value();
+
+  // Use a completion queue together with a generic client reader/writer to
+  // perform the method invocation.  All steps are done in serial, so just use
+  // consecutive integer values for completion queue tags (there's no need to use the tag values for
+  // correlation).
+  LOG(INFO) << "Invoke method " << method_name;
+  grpc::ClientContext ctx;
+  grpc::CompletionQueue cq;
+  std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call =
+      stub_->PrepareCall(&ctx, method_name, &cq);
+  call->StartCall(tag(1));
+
+  void* got_tag;
+  bool ok;
+  cq.Next(&got_tag, &ok);
+  if (!ok) {
+    LOG(ERROR) << "Failed to start gRPC method call";
+    return false;
+  }
+
+  // Send the un-encapsulated gRPC request (which is already serialized).
+  grpc::Slice slice(req_data.data(), req_data.size());
+  grpc::ByteBuffer send_bb(&slice, /*nslices=*/1);
+  call->Write(send_bb, tag(2));
+  cq.Next(&got_tag, &ok);
+  if (!ok) {
+    LOG(ERROR) << "Failed to send gRPC request";
+    return false;
+  }
+  call->WritesDone(tag(3));
+  cq.Next(&got_tag, &ok);
+  if (!ok) {
+    LOG(ERROR) << "Failed to close gRPC request";
+    return false;
+  }
+
+  // Loop round reading responses off the completion queue (which allows for
+  // server-streaming gRPC methods).
+  while (true) {
```

Comment:
Wow, thanks for figuring out all this stuff, I can imagine how painful it must have been, having done something similar for the generic server.

Reply:
These look like useful references to keep in the code for future maintainers.